src/electionguard/group.py

```diff
@@ -17,7 +19,57 @@
 from .constants import get_large_prime, get_small_prime, get_generator
 
 
-class BaseElement(ABC, int):
+class BaseElementMeta(type):
+    """Metaclass to cast return type of int functions to subclass type"""
+
+    def __new__(cls, name: str, bases: tuple, attrs: dict) -> BaseElementMeta:
+        obj = super().__new__(cls, name, bases, attrs)
+
+        for k, v in int.__dict__.items():
+            # do not override base existing (__hash__, __new__, __doc__, etc)
+            if k in attrs:
+                continue
+
+            # do not override functions meant to return `int`/`float`
+            if k in ["__int__", "__float__"]:
+                continue
+
+            # do not override, manually overriden methods
+            if k in ["__eq__", "__ne__", "__hash__"]:
+                continue
+
+            # override the return types to return subclass type
+            if getattr(v, "__objclass__", None) is int:
+                setattr(obj, k, obj.cast_as_subclass(v))
+
+        return obj
+
+    def cast_as_subclass(cls, obj: Any) -> Callable:
+        def convert(value: Any) -> Any:
+            if value.__class__ is int:
+                return cls(value)
+
+            if value.__class__ is float:
+                if value == int(value):
+                    return cls(int(value))
+
+                raise TypeError(f"{cls.__name__} does not support float values.")
+
+            if value is NotImplemented:
+                raise TypeError(
+                    f"{cls.__name__} only interacts with int, ElementModP, or ElementModQ"
+                )
+
+            return value
+
+        @wraps(obj)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            return convert(obj(*args, **kwargs))
+
+        return wrapper
+
+
+class BaseElement(int, metaclass=BaseElementMeta):
```

Comment:
We've hit `metaclass`. Oh noes. This isn't quite what I was thinking. Let's discuss. Reach out view email.